[BUGFIX] Rendre l'affichage de la date de naissance du certificat insensible aux fuseaux horaires (PIX-1824)

Giroteti opened the pull request:
## :unicorn: Problème
Le formatage de la date de naissance sur le certificat (front `mon-pix`) dépend du fuseau horaire dans lequel se trouve l'utilisateur. Par exemple la date "brute" `2020-01-01` sera affichée "1 janvier 2020" en France et "31 décembre 1999" à Tahiti. 
Le problème réside dans le fait que le `formatDate` du service `ember-intl` convertit à la volée une string en date à l'aide d'un `new Date(laDate)` sans spécifier de timezone. Le comportement par défaut utilise la timezone du navigateur. 

## :robot: Solution
Convertir explicitement la string en date avant d'appeler `formatDate` et spécifier également la timezone du formatage attendu.

## :rainbow: Remarques
RAS

## :100: Pour tester
En local :
- se connecter avec certif-success@example.net
- consulter le certificat
- constater l'affichage de la date de naissance "1 janvier 2000"
- modifier la timezone du navigateur (sous chrome : console de dev > trois petit points > more tools > sensors > location)
- recharger la page 
- constater que l'affichage de la date de naissance est toujours "1 janvier 2000" (avec cette PR, il aurait été "31 décembre 1999")

Giroteti committed:
Extract date formatting to a service

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2384.review.pix.fr
- Orga: https://orga-pr2384.review.pix.fr
- Certif: https://certif-pr2384.review.pix.fr
- Admin: https://admin-pr2384.review.pix.fr
- API: https://api-pr2384.review.pix.fr/api/

 Please check it out!

Anne-Gaelle-S commented on mon-pix/tests/unit/services/intl-date-formatter-test.js:
```diff
@@ -0,0 +1,36 @@
+    const dateString = '2020-31-12';
+
+    // when / then
+    expect(() => intlDateFormatter.formatDateStringToString(dateString))
+      .to.throw();
```
**suggestion (non bloquante)**
Pour chippoter 😈 : pourquoi avoir mis sur 2 lignes ? On aurait pu mettre ça en une seule je pense

Anne-Gaelle-S reviewed (approved):
GG pour avoir trouvé d'où le bug venait !! 🎉

jonathanperret commented:
> Le problème réside dans le fait que le `formatDate` du service `ember-intl` convertit à la volée une string en date à l'aide d'un `new Date(laDate)` sans spécifier de timezone. Le comportement par défaut utilise la timezone du navigateur.

Pour être précis et pour mémoire : dans ce cas précis (chaîne au format `YYYY-MM-DD`), `new Date()` [utilise systématiquement la timezone UTC](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#parameters). La dépendance à la timezone du navigateur intervient dans la conversion de cet objet `Date` en chaîne que réalise `ember-intl`.

Pour le formatage `ember-intl` [s'appuie sur l'objet natif `Intl.DateTimeFormat`](https://github.com/ember-intl/ember-intl/blob/ba72730c9f0522b6117ac0c584352a9d76bc8bc5/addon/-private/formatters/format-date.js#L37) qui va utiliser par défaut la timezone du navigateur.

Là est l'origine du problème : une chaîne du type `2000-01-02` est d'abord convertie en instant (nombre de millisecondes depuis le 1er janvier 1970) par `new Date()` en considérant l'instant où il était minuit à la date en question _dans la timezone UTC_, puis cet instant est représenté en texte en considérant la date et l'heure qu'il représente _dans la timezone du navigateur_.

Par exemple, `2000-01-02` interprété comme minuit UTC c'est `946771200000` millisecondes après le 1er janvier 1970. À cet instant-là : 

- à Londres (UTC en hiver) il était 00:00:00 le 2 janvier ;
- à Paris (UTC+1 en hiver) il était 01:00:00 le 2 janvier ;
- à La Réunion (UTC+4) il était 04:00:00 le 2 janvier ;
- aux Kiribati (UTC+14) il était 14:00:00 le 2 janvier ;
- en Guadeloupe (UTC-4) il était 20:00:00 le 1er janvier ;
- aux Îles Marquises (UTC-9:30) il était 09:30:00 le 1er janvier.

Comme on peut le voir, pour les timezones à l'est de Greenwich, pas de souci, la date affichée correspondra bien à ce qui est indiqué dans la donnée d'origine. En revanche dès qu'on est à l'ouest, un décalage apparaît.

La correction ici consiste à utiliser la même timezone pour convertir la date d'entrée en instant et pour convertir cet instant en texte. On pourrait choisir n'importe quelle timezone pour ça mais on a choisi UTC parce que c'est la plus simple.

jonathanperret commented on mon-pix/app/services/intl-date-formatter.js:
```diff
@@ -0,0 +1,23 @@
+function isDateStringFormatValid(dateString) {
+  const validFormat = 'YYYY-MM-DD';
+  const strictMode = true;
+  return moment(dateString, validFormat, strictMode).isValid();
```
Il y a peut-être moyen de faire un peu plus efficace en évitant de parser la date deux fois (une fois avec `moment` et une fois avec `new Date`).

Par exemple en utilisant directement le résultat du parsing fait par `moment` (probablement en utilisant `moment.utc` pour garantir une interprétation UTC, ce qui permet de ne pas avoir à ajouter la chaîne `T00:00:00Z`) pour le passer à `formatDate` ;

Ou alors, éviter la dépendance à `moment` et faire une vérification plus basique comme `/^\d\d\d\d-\d\d-\d\d$/.test(dateString)` : ce n'est pas très strict mais l'idée n'est pas de se protéger d'une saisie arbitraire de données (l'API renvoie des dates correctes), plutôt d'une mauvaise utilisation de la fonction où quelqu'un passerait un objet `Date` ou une chaîne contenant une heure.

Giroteti committed:
Force certif and deliveredAt date to Europe/Paris

The date on which the certification test was taken should be displayed using the timezone of the certification center in which it was held. As of today we do not have the timezone of the certification centers, the vast mojority of them being in France we choose to display those date in Europe/Paris time for the time being (pun intended).
